torch/profiler/_utils.py

```diff
@@ -18,32 +42,214 @@ def __eq__(self, other):
+            overlap_end = min(self.event.end_time_ns, interval.end)
 
-@dataclass
-class EventMetrics:
-    self_time_ns: int = 0
+            if overlap_start < overlap_end:
+                overlap_time += overlap_end - overlap_start
+        return overlap_time
+
+
+class BasicEvaluation:
+
+    def __init__(self, prof: profile):
+        self.profile = prof
+        self.metrics: Dict[EventKey, EventMetrics] = {}
+        self.compute_self_time()
+        self.qd_list = self.compute_idle_time_and_qd()
+
+    def compute_self_time(self):
+        '''
+        Computes event's self time(total time - time in child ops).
+        '''
+        assert (self.profile.kineto_results is not None)
+        stack = deque(self.profile.kineto_results.experimental_event_tree())
+
+        # standard iterating dfs
+        while stack:
+            curr_event = stack.pop()
+            self_time = curr_event.duration_time_ns
+            for child_event in curr_event.children:
+                self_time -= child_event.duration_time_ns
+                stack.append(child_event)
+
+            assert EventKey(
+                curr_event
+            ) not in self.metrics, f"Duplicate id: {curr_event.id}, {curr_event.name()}"
+            self.metrics[EventKey(curr_event)] = EventMetrics(
+                self_time_ns=self_time)
+            self.metrics[EventKey(
+                curr_event)].duration_time_ns = curr_event.duration_time_ns
+
+    def compute_idle_time_and_qd(self):
+        '''
+        Computes event's idle time. Idle time is defined as the time when the CUDA kernel queue depth is 0.
+        It also return a Time series of the queue depth data.
+        qd = cuda kernel queue depth
+        '''
+        assert (self.profile.kineto_results is not None)
+        event_list = self.profile.kineto_results.events()
+
+        def is_cuda_launch_kernel(e):
+            # TODO: find a better way to identify cudaLaunchKernel
+            return e.name() == "cudaLaunchKernel"
 
+        def is_cuda_kernel(e):
+            # TODO: find a better way to identify CUDA Kernel
+            return e.device_type() == DeviceType.CUDA and "mem" not in e.name(
+            ).lower()
 
-def compute_self_time(prof: profile, metrics: Dict[EventKey, EventMetrics]):
-    '''
-    Computes event's self time(total time - time in child ops).
+        # Record All the idle intervals
+        curr_qd = 0
+        idle_interval = []
+        qd_list = []
+        idle_start = 0
+        cuda_kernel_events = [
+            event for event in event_list
+            if is_cuda_launch_kernel(event) or is_cuda_kernel(event)
+        ]
+        cuda_kernel_events.sort(key=lambda e: e.start_us())
+
+        for curr_event, next_event in zip(cuda_kernel_events,
+                                          cuda_kernel_events[1:]):
+            if (is_cuda_launch_kernel(curr_event)):
+                if (curr_qd == 0):
```

Comment:
This method of counting strikes me as rather brittle. If you have a single launch where we don't capture the kernel then `curr_qd` will never go to zero. And by contrast, if there are more kernels than launches (recall that kernels can call other kernels in CUDA) curr_qd will be negative and all sorts of craziness will happen.

A potentially more robust solution would be to keep track of blocks of kernels (`List[List[_KinetoEvent]]`) and make a new block when a kernel launch had a start time greater than the end time of the last event in the current block.

Reply:
I am not quite getting your solution here, could you elaborate on that a little bit?

Reply:
After thinking a bit I decided the nested lists probably aren't needed. But the general procedure is the same.

Suppose we partition events into kernel launches and kernels, and maintain a mapping from kernel to launch. (If not directly assignable by correlation ID, check previous kernels until we find one that is.) We can then walk the cpu events and get the gpu kernel currently executing when the event starts (if any) by advancing an index into the GPU kernels. We know the index of the kernel launched by the current event (or the last launch if the current event is not a cuda kernel launch), and the difference in those indices is the queue depth at that op.